l2-contracts/contracts/bridge/L2SharedBridgeLegacy.sol

```diff
@@ -2,26 +2,142 @@
+            revert EmptyBytes32();
+        }
+
+        if (_aliasedOwner == address(0)) {
+            revert EmptyAddress();
+        }
+
+        l1SharedBridge = _l1SharedBridge;
+
+        if (block.chainid != ERA_CHAIN_ID) {
+            address l2StandardToken = address(new L2StandardERC20{salt: bytes32(0)}());
+            l2TokenBeacon = new UpgradeableBeacon{salt: bytes32(0)}(l2StandardToken);
+            l2TokenProxyBytecodeHash = _l2TokenProxyBytecodeHash;
+            l2TokenBeacon.transferOwnership(_aliasedOwner);
+        } else {
+            if (_l1Bridge == address(0)) {
+                revert EmptyAddress();
+            }
+            l1Bridge = _l1Bridge;
+            // l2StandardToken and l2TokenBeacon are already deployed on ERA, and stored in the proxy
+        }
+    }
+
+    /// @notice Initiates a withdrawal by burning funds on the contract and sending the message to L1
+    /// where tokens would be unlocked
+    /// @param _l1Receiver The account address that should receive funds on L1
+    /// @param _l2Token The L2 token address which is withdrawn
+    /// @param _amount The total amount of tokens to be withdrawn
+    function withdraw(address _l1Receiver, address _l2Token, uint256 _amount) external override {
+        if (_amount == 0) {
+            revert AmountMustBeGreaterThanZero();
+        }
+        L2_ASSET_ROUTER.withdrawLegacyBridge(_l1Receiver, _l2Token, _amount, msg.sender);
+    }
+
+    /// @return Address of an L2 token counterpart
+    function l2TokenAddress(address _l1Token) public view override returns (address) {
+        address token = L2_NATIVE_TOKEN_VAULT.l2TokenAddress(_l1Token);
+        if (token != address(0)) {
+            return token;
+        }
+        return calculateCreate2TokenAddress(_l1Token);
+    }
+
+    /// @notice Calculates L2 wrapped token address given the currently stored beacon proxy bytecode hash and beacon address.
+    /// @param _l1Token The address of token on L1.
+    /// @return Address of an L2 token counterpart.
+    function calculateCreate2TokenAddress(address _l1Token) internal view returns (address) {
+        bytes32 constructorInputHash = keccak256(abi.encode(address(l2TokenBeacon), ""));
+        bytes32 salt = _getCreate2Salt(_l1Token);
+        return
+            L2ContractHelper.computeCreate2Address(address(this), salt, l2TokenProxyBytecodeHash, constructorInputHash);
+    }
+
+    /// @dev Convert the L1 token address to the create2 salt of deployed L2 token
+    function _getCreate2Salt(address _l1Token) internal pure returns (bytes32 salt) {
+        salt = bytes32(uint256(uint160(_l1Token)));
+    }
+
+    /// @dev Deploy the beacon proxy for the L2 token, while using ContractDeployer system contract.
+    /// @dev This function uses raw call to ContractDeployer to make sure that exactly `l2TokenProxyBytecodeHash` is used
+    /// for the code of the proxy.
+    function deployBeaconProxy(bytes32 salt) external returns (address proxy) {
```

Comment:
please add access control. someone from outside could deploy a beacon proxy with a salt equal to an L1 token address and make all deployments coming from NTV fail